### ISSUE TITLE
[BUG] Sortable select fields don't remain sorted.

### ISSUE THREAD

**Issue by graham73may** [bolt/bolt#6291]
Details
-------

 - Relevant Bolt Version: 3.2.6
 - Install type: Composer install
 - PHP version: 7.0.9
 - Used web server: Apache
 - For UX/UI issues: Chrome latest

Reproduction
------------
​
 1. **Bug summary**: 
 
Sortable select fields change to being alphabetically sorted on save. You can re-order the items after the page has been saved. 

 2. **Specifics**:

 3. **Steps to reproduce**:

Add a sortable select field to `contenttypes.yml`

```
        authors:
            group: "Meta"
            label: "Authors"
            multiple: true
            sortable: true
            autocomplete: true
            type: select
            values: people/title
```

Load up the contenttype your fields are on, add some items in a specific order. Press save. 

 4. **Expected result**: 
Expect to see the items saved in the order they appeared in the field. 

 5. **Actual result**: 
The items in the field become alphabetically sorted. 

It is however possible to sort the items once you have saved the page once and re-drag the items.

**Comment by rossriley** [comment]
Looks like a fix for another bug has impacted on this, but it may be version specific since I'm not sure what point release this went in.

In #6096 the call that sorts the values based on the selected options (the unique twig helper was originally brought over from the extension) got removed so I'm pretty sure that will stop the sortable working altogether.....
https://github.com/bolt/bolt/pull/6096/files#diff-b0666fa264092eada1a5d5a18c75a4faL45

I can't get my head around this block:

```
 foreach ($merged as $val) {
     $compiled[$val[0]] = $val;
 }
```

I'm sure it must have worked at some point since it's been part of the extension from the beginning but you are right, as things stand now it seems to be pulling the first character of a string out, where I'm sure it must have previously been pulling out the index position instead.

So rough guess is that things in the underlying selectvalues have moved around somehow. Crux is it will need reworking and what you pasted in Slack is probably a good place to start.

**Comment by mrenigma** [comment]
@rossriley I think that was one of our bugs which resulted in that pr :smile: 

I thought I would put what I found on here as well as slack so it doesn't get lost.

> I've come a cross an interesting issue and I don't know if it's intended. It looks like a standard sortable multiple select field with an `id,title` value tries to use the `ArrayHandler` unique twig function and assigns `$val[0]` to be the key even when `$val` is a string.
>
> Resulting in the following (twig variables dumped in order: `newvals`, `values`, `options`, `selection`):

![pasted image at 2017_01_29 09_45 pm](https://cloud.githubusercontent.com/assets/1901969/22415796/f296e9d8-e6c1-11e6-8274-49170556463e.png)

> So "Beatum, inquit." should have a options value of `5` but because the custom unique function references a string it ends up being `B`... And therefore nothing is selected

And as ross mentioned the pasted twig code from slack:

```twig
{% if field.sortable|default(false) %}
    {% set ordered_options = [] %}
    {% set dontorder_options = [] %}
    {% set assigned = [] %}
    {% for selected in selection %}
        {% for option in options %}
            {% if option.value not in assigned %}
                {% if option.value == selected %}
                    {% set ordered_options = ordered_options|merge([option]) %}
                    {% set assigned = assigned|merge([option.value]) %}
                {% elseif option.selected != true %}
                    {% set dontorder_options = dontorder_options|merge([option]) %}
                    {% set assigned = assigned|merge([option.value]) %}
                {% endif %}
            {% endif %}
        {% endfor %}
    {% endfor %}

    {% set options = ordered_options|merge(dontorder_options) %}
{% endif %}
```

It's worth noting I realised a potential edge-case last night so the code got slightly longer as a result -before if the selects weren't sequential in the options array, it would end up putting selected items in the `dontorder_options` array.

**Comment by GawainLynch** [comment]
Fixed in #6354 & #6355
